pkg/morph/deploy/notary.go

```diff
@@ -0,0 +1,872 @@
+	}
+
+	if len(b) != sharedTransactionDataLen {
+		return fmt.Errorf("invalid/unsupported length of shared transaction data: expected %d, got %d",
+			sharedTransactionDataLen, len(b))
+	}
+
+	x.sender, err = util.Uint160DecodeBytesBE(b[:util.Uint160Size])
+	if err != nil {
+		return fmt.Errorf("decode sender account binary: %w", err)
+	}
+
+	x.validUntilBlock = binary.BigEndian.Uint32(b[util.Uint160Size:])
+	x.nonce = binary.BigEndian.Uint32(b[util.Uint160Size+4:])
+
+	return nil
+}
+
+const (
+	sharedTransactionDataLen         = util.Uint160Size + 4 + 4
+	sharedTransactionDataChecksumLen = 4
+)
+
+// unshiftChecksum prepends given payload with first 4 bytes of the
+// sharedTransactionData SHA-256 checksum. Inverse operation to shiftChecksum.
+func (x sharedTransactionData) unshiftChecksum(data []byte) []byte {
+	h := sha256.Sum256(x.bytes())
+	return append(h[:sharedTransactionDataChecksumLen], data...)
+}
+
+// shiftChecksum matches checksum of the sharedTransactionData and returns
+// payload. Inverse operation to unshiftChecksum.
+func (x sharedTransactionData) shiftChecksum(data []byte) (bool, []byte) {
+	if len(data) < sharedTransactionDataChecksumLen {
+		return false, data
+	}
+
+	h := sha256.Sum256(x.bytes())
+	if !bytes.HasPrefix(data, h[:sharedTransactionDataChecksumLen]) {
+		return false, nil
+	}
+
+	return true, data[sharedTransactionDataChecksumLen:]
+}
+
+// sharedTxDataMatches checks if given transaction is constructed using provided
+// shared parameters.
+func sharedTxDataMatches(tx *transaction.Transaction, sharedTxData sharedTransactionData) bool {
+	return sharedTxData.nonce == tx.Nonce &&
+		sharedTxData.validUntilBlock == tx.ValidUntilBlock &&
+		len(tx.Signers) > 0 && tx.Signers[0].Account.Equals(sharedTxData.sender)
+}
+
+// makeUnsignedDesignateCommitteeNotaryTx constructs unsigned transaction that
+// designates Notary role to the specified committee members using shared
+// parameters.
+func makeUnsignedDesignateCommitteeNotaryTx(roleContract *rolemgmt.Contract, committee keys.PublicKeys, sharedTxData sharedTransactionData) (*transaction.Transaction, error) {
+	committeeMultiSigAcc, err := notary.FakeMultisigAccount(smartcontract.GetMajorityHonestNodeCount(len(committee)), committee)
+	if err != nil {
+		return nil, fmt.Errorf("init committee multi-sig account: %w", err)
+	}
+
+	tx, err := roleContract.DesignateAsRoleUnsigned(noderoles.P2PNotary, committee)
+	if err != nil {
+		return nil, err
+	}
+
+	tx.ValidUntilBlock = sharedTxData.validUntilBlock
+	tx.Nonce = sharedTxData.nonce
+	tx.Signers = []transaction.Signer{
+		{
+			Account: sharedTxData.sender,
+			Scopes:  transaction.None,
+		},
+		{
+			Account: committeeMultiSigAcc.ScriptHash(),
+			Scopes:  transaction.CustomGroups | transaction.CalledByEntry,
+		},
+	}
+
+	// FIXME: w/o this transaction fails with 'invalid signature' error due to lack of GAS for some period
+	// (later succeeds when there is more GAS)
```

Comment:
`invalid signature` is exactly that, an invalid signature. Insufficient network fee should be `vm execution has failed`. And you network fee should be precise since you have all signers known and not changed between estimating/verifying (sender hash doesn't count, he has the same script cost anyway).

Reply:
double-checked, now there is no problem, dropped